src/epoch/participant.rs

```diff
@@ -0,0 +1,176 @@
+                // As it isn't already at the global epoch count, set it.
+                self.epoch.store(global_epoch, atomic::Ordering::Relaxed);
+
+                // Then collect the garbage.
+                unsafe {
+                    (*self.garbage.get()).collect();
+                }
+            }
+
+            true
+        }
+    }
+
+    /// Exit the current (nested) critical section.
+    pub fn exit(&self) {
+        // Decrement the counter.
+        self.in_critical.fetch_sub(1, atomic::Ordering::SeqCst);
+    }
+
+    /// Begin the reclamation process for a piece of data.
+    pub unsafe fn reclaim<T>(&self, data: *mut T) {
+        // Insert it into the garbage list.
+        (*self.garbage.get()).insert(data);
+    }
+
+    /// Attempt to collect garbage by moving the global epoch forward.
+    ///
+    /// If it failed, it returns `Err(())`.
+    pub fn try_collect(&self, guard: &Guard) -> Result<(), ()> {
+        // Load the global epoch counter.
+        let cur_epoch = global::EPOCH.epoch.load(atomic::Ordering::SeqCst);
+
+        // Go over the participants to ensure that none are critical.
+        for p in global::EPOCH.participants.iter(guard) {
+            if p.is_critical() && p.epoch.load(atomic::Ordering::Relaxed) != cur_epoch {
+                // A participant was in critical state, and can thus not be GC'd at this moment.
+                return Err(());
+            }
+        }
+
+        let new_epoch = cur_epoch.wrapping_add(1);
+
+        // This fence is important to avoid reordering of the CAS with any of the previous loads.
+        // TODO: Is this really necessary?
+        atomic::fence(atomic::Ordering::Acquire);
+        // CAS with the new epoch. The reason we do CAS and not a simple store is that we want to
+        // solve the ABA problem, as another thread could have updated the counter, causing us to
+        // reverting its changes. Hence, we must ensure that this is NOT the case.
+        if global::EPOCH.epoch.compare_and_swap(cur_epoch, new_epoch, atomic::Ordering::SeqCst) != cur_epoch {
+            // Another thread touched it, so the collection failed.
+            return Err(());
+        }
+
+        // If you've come so far, everything is right for collecting the garbage, so let's do it.
+        unsafe {
+            (*self.garbage.get()).collect();
+            global::EPOCH.garbage.collect(new_epoch.wrapping_add(1))
+        }
+
+        // Update the epoch counter. As the epoch was collected, the ABA problem shouldn't be an
+        // issue here.
+        // TODO: Is this really correct?
+        self.epoch.store(new_epoch, atomic::Ordering::Release);
+
+        Ok(())
+    }
+
+    /// Move the current thread-local garbage into the global garbage bags.
+    pub fn migrate_garbage(&self) {
+        // Load the current epoch counter.
+        let cur_epoch = self.epoch.load(atomic::Ordering::Relaxed);
+        // Replace the local garbage set with an empty set.
```

Comment:
You could have this comment or instead do something like.

~~~
let empty = garbage::Local::default();
let local = unsafe { mem::replace(&must *self.garbage.get(), empty) };
~~~